cmd/kubeadm/app/cmd/reset.go

```diff
@@ -21,145 +21,207 @@ import (
+		fmt.Printf("[reset] Failed to unmount directories in /var/lib/kubelet: %s\n", string(umountOutputBytes))
 	}
 
-	dirsToClean := []string{"/var/lib/kubelet"}
+	// Remove contents from the config and pki directories
+	resetConfigDir(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmapi.GlobalEnvParams.HostPKIPath)
+
+	dirsToClean := []string{"/var/lib/kubelet", "/etc/cni/net.d"}
 
 	// Only clear etcd data when the etcd manifest is found. In case it is not found, we must assume that the user
 	// provided external etcd endpoints. In that case, it is his own responsibility to reset etcd
 	if _, err := os.Stat("/etc/kubernetes/manifests/etcd.json"); os.IsNotExist(err) {
 		dirsToClean = append(dirsToClean, "/var/lib/etcd")
+	} else {
+		fmt.Printf("[reset] No etcd manifest found in %q, assuming external etcd.\n", "/etc/kubernetes/manifests/etcd.json")
 	}
 
-	resetConfigDir("/etc/kubernetes/")
-
-	fmt.Printf("Deleting contents of stateful directories: %v\n", dirsToClean)
+	fmt.Printf("[reset] Deleting contents of stateful directories: %v\n", dirsToClean)
 	for _, dir := range dirsToClean {
 		cleanDir(dir)
 	}
 
 	dockerCheck := preflight.ServiceCheck{Service: "docker"}
 	if warnings, errors := dockerCheck.Check(); len(warnings) == 0 && len(errors) == 0 {
-		fmt.Println("Stopping all running docker containers...")
-		if err := exec.Command("sh", "-c", "docker ps | grep 'k8s_' | awk '{print $1}' | xargs docker rm --force --volumes").Run(); err != nil {
-			fmt.Println("failed to stop the running containers")
+		fmt.Println("[reset] Stopping all running docker containers...")
+		if err := exec.Command("sh", "-c", "docker ps | grep 'k8s_' | awk '{print $1}' | xargs -r docker rm --force --volumes").Run(); err != nil {
+			fmt.Println("[reset] Failed to stop the running containers")
 		}
 	} else {
-		fmt.Println("docker doesn't seem to be running, skipping the removal of kubernetes containers")
+		fmt.Println("[reset] docker doesn't seem to be running, skipping the removal of running kubernetes containers")
+	}
+
+	return nil
+}
+
+func drainAndRemoveNode(removeNode bool) error {
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("failed to detect node hostname")
+	}
+	hostname = strings.ToLower(hostname)
+
+	// TODO: Use the "native" k8s client for this once we're confident the versioned is working
+	kubeConfigPath := path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "kubelet.conf")
+
+	getNodesCmd := fmt.Sprintf("kubectl --kubeconfig %s get nodes | grep %s", kubeConfigPath, hostname)
+	output, err := exec.Command("sh", "-c", getNodesCmd).Output()
+	if err != nil || len(output) == 0 {
+		// kubeadm shouldn't drain and/or remove the node when it doesn't exist anymore
+		return nil
+	}
+
+	fmt.Printf("[reset] Draining node: %q\n", hostname)
+
+	output, err = exec.Command("kubectl", "--kubeconfig", kubeConfigPath, "drain", hostname, "--delete-local-data", "--force", "--ignore-daemonsets").Output()
```

Comment:
It works on nodes as well, otherwise I wouldn't have added it here...
The `/etc/kubernetes/kubelet.conf` file exists everywhere when using kubeadm, and kubectl is installed everywhere, so when going the opinionated route with debs and rpms, it just "should work (tm)" :smile: